hugo/content/docs/recipes/performance/caches.md

```diff
@@ -0,0 +1,215 @@
+    Egmont: ['Asterix', 'Obelix']
+};
+```
+
+### Define the computation service
+
+For our service we define an interface:
+
+```typescript
+export interface InferPublisherService {
+    inferPublisher(person: Person): KnownPublisher|undefined;
+}
+```
+
+Now we implement the service:
+
+```typescript
+class UncachedInferPublisherService implements InferPublisherService {
+    inferPublisher(person: Person): KnownPublisher|undefined {
+        for (const [publisher, persons] of Object.entries(KnownPersonNames)) {
+            if(persons.includes(person.name)) {
+                return publisher as KnownPublisher;
+            }
+        }
+        return undefined;
+    }
+}
+```
+
+### Add a cache
+
+Now we want to cache the results of the `inferPublisher` method. We can use the `DocumentCache` for this. We will reuse the uncached service as base class and override the `inferPublisher` method:
+
+```typescript
+export class CachedInferPublisherService extends UncachedInferPublisherService {
+    private readonly cache: DocumentCache<Person, KnownPublisher|undefined>;
+    constructor(services: HelloWorldServices) {
+        super();
+        this.cache = new DocumentCache(services.shared);
+    }
+    override inferPublisher(person: Person): KnownPublisher|undefined {
+        const documentUri = AstUtils.getDocument(person).uri;
+        //get cache entry for the documentUri and the person
+        //if it does not exist, calculate the value and store it
+        return this.cache.get(documentUri, person, () => super.inferPublisher(person));
+    }
+}
+```
+
+### Use the service
+
+To use this service, let's create a validator that checks if the publisher of a person is known. Go to the `hello-world-validator.ts` file and add the following code:
+
+```typescript
+import type { ValidationAcceptor, ValidationChecks } from 'langium';
+import type { HelloWorldAstType, Person } from './generated/ast.js';
+import type { HelloWorldServices } from './hello-world-module.js';
+import { InferPublisherService } from './infer-publisher-service.js';
+
+/**
+ * Register custom validation checks.
+ */
+export function registerValidationChecks(services: HelloWorldServices) {
+    const registry = services.validation.ValidationRegistry;
+    const validator = services.validation.HelloWorldValidator;
+    const checks: ValidationChecks<HelloWorldAstType> = {
+        Person: validator.checkPersonIsFromKnownPublisher
+    };
+    registry.register(checks, validator);
+}
+
+/**
+ * Implementation of custom validations.
+ */
+export class HelloWorldValidator {
+    private readonly inferPublisherService: InferPublisherService;
+
+    constructor(services: HelloWorldServices) {
+        this.inferPublisherService = services.utilities.inferPublisherService;
+    }
+
+    checkPersonIsFromKnownPublisher(person: Person, accept: ValidationAcceptor): void {
+        if(this.inferPublisherService.inferPublisher(person) === undefined) {
```

Comment:
Formatting. Also comparisons with `undefined` are unusual in TS/JS. Just use the falsyness check
```suggestion
        if (!this.inferPublisherService.inferPublisher(person)) {
```